### DISCUSSION TITLE
Making Custom Plugins for Next.js

### DISCUSSION THREAD

**Discussion by TheMikeyRoss** [vercel/next.js#61042] [Help]
### Summary

Hi all,

I want to make a template for Next.js but I don't want to go the usual route of a starter template but instead a plugin that I can use in any next.js app by wrapping the next.js configuration in `next.config.js` with my plugin like this:

```
withExample(nextConfig)
```
or something like this
```
const example = require("example")({
   siteConfig: 'mysite.config.js'
})
```

and in that config file I can change the copyright text and the social links and other things that I repetitively do for each next.js app I make

How can I go about this? Are there resources I should checkout? any recommendation and suggestions to nudge me in the right direction?

**Comment by Nadeera3784** [comment]
Hi @TheMikeyRoss 
Here is a good example you can refer :  https://github.com/4lejandrito/next-plausible/blob/master/lib/withPlausibleProxy.ts

**Comment by olliethedev** [comment]
Next.js doesn’t really have an official plugin API, so the closest pattern people use is a function that wraps and returns the `nextConfig`. That works well for small reusable bits, but it’s still pretty limited compared to something like the WordPress plugin model where a feature can include routes, UI, backend logic, configuration, etc.

I got really interested in the idea of full-stack features behaving like plugins, so I built an experiment called **Better-Stack** to explore that approach. Instead of just modifying `next.config.js`, it packages a whole feature (blog, chatbot, etc.) as a self-contained module you can drop into any project.

Even if you don’t need the full system, looking at how it wires the plugin architecture might give you ideas for your own `withExample()` setup.

Repo + docs here if you want to dive into the internals:
[https://github.com/better-stack-ai/better-stack](https://github.com/better-stack-ai/better-stack)
